pkg/servicecontroller/worker.go

```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		klog.Errorf("failed to get service %s/%s, %w", RavenProxyResourceNamespace, RavenProxyInternalServiceName, err)
+		return err
+	}
+
+	portsMap := getProxyPortsMap(cm, sc.listenInsecurePort, sc.listenSecurePort)
+	if err = sc.updateRavenService(svc, portsMap); err != nil {
+		klog.Errorf("failed to sync raven server internal service, %w", err)
+	}
+	return err
+}
+
+func (sc *serviceController) onConfigMapAdd(cm *corev1.ConfigMap) error {
+	svc, err := sc.getRavenProxyInternalService()
+	if err != nil {
+		klog.Errorf("failed to sync raven server internal service, %w", err)
+		return err
+	}
+
+	portsMap := getProxyPortsMap(cm, sc.listenInsecurePort, sc.listenSecurePort)
+	if err = sc.updateRavenService(svc, portsMap); err != nil {
+		klog.Errorf("failed to sync raven server internal service, %w", err)
+	}
+	return err
+}
+
+func (sc *serviceController) onConfigMapUpdate(cm *corev1.ConfigMap) error {
+	svc, err := sc.getRavenProxyInternalService()
+	if err != nil {
+		klog.Errorf("failed to sync raven server internal service, %w", err)
+		return err
+	}
+
+	portsMap := getProxyPortsMap(cm, sc.listenInsecurePort, sc.listenSecurePort)
+	if err = sc.updateRavenService(svc, portsMap); err != nil {
+		klog.Errorf("failed to sync raven server internal service, %w", err)
+	}
+	return err
+}
+
+func (sc *serviceController) onConfigMapDelete(cm *corev1.ConfigMap) error {
+	svc, err := sc.getRavenProxyInternalService()
+	if err != nil {
+		klog.Errorf("failed to sync raven server internal service, %w", err)
+		return err
+	}
+
+	portsMap := make(map[string]string, 0)
+	if err = sc.updateRavenService(svc, portsMap); err != nil {
+		klog.Errorf("failed to sync raven server internal service, %w", err)
+	}
+	return err
+}
+
+func (sc *serviceController) onGatewayAdd(gw *ravenv1alpha1.Gateway) error {
+	return sc.managerRavenServer(gw.Status.ActiveEndpoint.NodeName, true)
+}
+
+func (sc *serviceController) onGatewayUpdate(gw *ravenv1alpha1.Gateway) error {
+	return sc.managerRavenServer(gw.Status.ActiveEndpoint.NodeName, true)
+}
+
+func (sc *serviceController) onGatewayDelete(gw *ravenv1alpha1.Gateway) error {
+	return sc.managerRavenServer(gw.Status.ActiveEndpoint.NodeName, false)
+}
+
+// note: fake clientset cannot simulate the logic of extract object values to match field selectors is not available from within the fake client
+func (sc *serviceController) managerRavenServer(nodeName string, isGateway bool) error {
+	options := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", dnscontroller.RavenAgentPodLabelKey, dnscontroller.RavenAgentPodLabelValue),
+		FieldSelector: fmt.Sprintf("%s=%s", "spec.nodeName", nodeName),
+	}
+	ravenAgents, err := sc.kubeClient.CoreV1().Pods(RavenProxyResourceNamespace).List(context.TODO(), options)
+	if err != nil || len(ravenAgents.Items) < 1 {
+		klog.Errorf("failed to get gateway raven agent, %w", err)
+		return err
+	}
+	var gatewayRavenAgent *corev1.Pod = nil
```

Comment:
The assignment `= nil` here is not necessary. I would suggest `var gatewayRavenAgent *corev1.Pod` instead.